src/Rector/Namespace_/FactoryDefinitionRector.php

```diff
@@ -0,0 +1,228 @@
+            }
+
+            $this->processFactoryConfiguration($factories[$name->toString()], $stmt->expr);
+
+            unset($node->stmts[$key]);
+        }
+
+        foreach ($factories as $factory) {
+            $node->stmts[] = $factory;
+        }
+
+        return $node;
+    }
+
+    private function isReAssignFactoryVariable(Expr $expr): bool
+    {
+        if (! $expr instanceof Assign) {
+            return false;
+        }
+
+        return $this->isName($expr->var, 'factory');
+    }
+
+    private function shouldSkipExpression(MethodCall $methodCall): bool
+    {
+        if (! $methodCall->args[0]->value instanceof ClassConstFetch) {
+            return true;
+        }
+
+        return ! $this->isNames($methodCall->name, ['define', 'state', 'afterMaking', 'afterCreating']);
+    }
+
+    private function getNameFromClassConstFetch(Expr $expr): ?Name
+    {
+        if (! $expr instanceof ClassConstFetch) {
+            return null;
+        }
+
+        if ($expr->class instanceof Expr) {
+            return null;
+        }
+
+        return $expr->class;
+    }
+
+    private function createFactory(string $name, Expr $expr): Class_
+    {
+        return $this->modelFactoryNodeFactory->createEmptyFactory($name, $expr);
+    }
+
+    private function processFactoryConfiguration(Class_ $class, MethodCall $methodCall): void
+    {
+        if (! $this->isName($methodCall->var, 'factory')) {
+            return;
+        }
+
+        if ($this->isName($methodCall->name, 'define')) {
+            $this->addDefinition($class, $methodCall);
+        }
+
+        if ($this->isName($methodCall->name, 'state')) {
+            $this->addState($class, $methodCall);
+        }
+
+        if (! $this->isNames($methodCall->name, ['afterMaking', 'afterCreating'])) {
+            return;
+        }
+
+        $name = $this->getName($methodCall->name);
+        if ($name === null) {
+            return;
+        }
+
+        $this->appendAfterCalling($class, $methodCall, $name);
+    }
+
+    private function addDefinition(Class_ $class, MethodCall $methodCall): void
+    {
+        if (count($methodCall->args) !== 2) {
+            return;
+        }
+
+        $callback = $methodCall->args[1]->value;
+        if (! $callback instanceof Closure) {
+            return;
+        }
+
+        $class->stmts[] = $this->modelFactoryNodeFactory->createDefinition($callback);
+    }
+
+    private function addState(Class_ $class, MethodCall $methodCall): void
+    {
+        if (count($methodCall->args) !== 3) {
+            return;
+        }
+
+        $class->stmts[] = $this->modelFactoryNodeFactory->createStateMethod($methodCall);
+    }
+
+    private function appendAfterCalling(Class_ $class, MethodCall $methodCall, string $name): void
+    {
+        if (count($methodCall->args) !== 2) {
+            return;
+        }
+
+        $closure = $methodCall->args[1]->value;
```

Comment:
need to have `isset($methodCall->args[1])` check to skip when index removed by other rules.